sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts

```diff
@@ -333,6 +334,107 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
     return true;
   }
 
+  /**
+   * Updates the last reported assessment for the current assessment.
+   * @param assessmentId related assessment id.
+   * @param entityManager entity manager to be part of the transaction.
+   * @returns updated assessment.
+   */
+  async updateLastReportedAssessment(
+    assessmentId: number,
+    entityManager: EntityManager,
+  ): Promise<UpdateResult | undefined> {
+    const assessment = await this.repo.findOne({
+      select: {
+        id: true,
+        offering: {
+          studyStartDate: true,
+          studyEndDate: true,
+        },
+        application: {
+          id: true,
+        },
+      },
+      relations: {
+        application: true,
+        offering: true,
+      },
+      where: { id: assessmentId },
+    });
+    const studentAssessmentRepo =
+      entityManager.getRepository(StudentAssessment);
+    const auditUser = this.systemUsersService.systemUser;
+    const lastReportedAssessment = await this.getLastReportedAssessment(
+      assessment.application.id,
+      entityManager,
+    );
+    if (
+      !lastReportedAssessment ||
+      (assessment.offering.studyStartDate ===
+        lastReportedAssessment.offering.studyStartDate &&
+        assessment.offering.studyEndDate ===
+          lastReportedAssessment.offering.studyEndDate)
+    ) {
+      return;
+    }
+    return studentAssessmentRepo.update(
+      {
+        id: assessmentId,
+      },
+      {
+        previousDateChangedReportedAssessment: lastReportedAssessment,
+        modifier: auditUser,
+      },
+    );
+  }
+
+  /**
+   * Gets the last reported assessment.
+   * @param applicationId current assessment application id.
+   * @param entityManager entity manager to be part of the transaction.
+   * @returns last reported assessment.
+   */
+  private async getLastReportedAssessment(
+    applicationId: number,
+    entityManager: EntityManager,
+  ): Promise<StudentAssessment> {
+    const studentAssessmentRepo =
+      entityManager.getRepository(StudentAssessment);
+    const lastApplicationChangeReportedAssessment =
+      await studentAssessmentRepo.findOne({
+        select: {
+          id: true,
+          assessmentDate: true,
+          offering: { id: true, studyStartDate: true, studyEndDate: true },
+        },
+        relations: { offering: true },
+        where: {
+          application: { id: applicationId },
+          previousDateChangedReportedAssessment: Not(IsNull()),
+          reportedDate: Not(IsNull()),
+        },
+        order: { assessmentDate: "DESC" },
+      });
```

Comment:
Are we not considering the limit 1 option ??

Reply:
I checked the query. It applies the limit by itself because of `findOne()`. The idea was to lookup that option if the current method is not applying the `limit`. Below is the generated query:

```sql
SELECT
		"StudentAssessment"."id" AS "StudentAssessment_id",
		"StudentAssessment"."assessment_date" AS "StudentAssessment_assessment_date",
		"StudentAssessment__StudentAssessment_offering"."id" AS "StudentAssessment__StudentAssessment_offering_id",
		"StudentAssessment__StudentAssessment_offering"."study_start_date" AS "StudentAssessment__StudentAssessment_offering_study_start_date",
		"StudentAssessment__StudentAssessment_offering"."study_end_date" AS "StudentAssessment__StudentAssessment_offering_study_end_date"
FROM
		"sims"."student_assessments" "StudentAssessment"
LEFT JOIN "sims"."education_programs_offerings" "StudentAssessment__StudentAssessment_offering" ON
		"StudentAssessment__StudentAssessment_offering"."id" = "StudentAssessment"."offering_id"
LEFT JOIN "sims"."applications" "StudentAssessment__StudentAssessment_application" ON
		"StudentAssessment__StudentAssessment_application"."id" = "StudentAssessment"."application_id"
LEFT JOIN "sims"."disbursement_schedules" "StudentAssessment__StudentAssessment_disbursementSchedules" ON
		"StudentAssessment__StudentAssessment_disbursementSchedules"."student_assessment_id" = "StudentAssessment"."id"
WHERE
		"StudentAssessment__StudentAssessment_application"."id" = $1
ORDER BY
	"StudentAssessment"."assessment_date" DESC,
	"StudentAssessment"."id" ASC
LIMIT 1;
```